File: frontend/src/services/schedule_service.ts
/**
 * Schedule Service — API сервис для работы с расписанием.
 * 
 * Работает с реальным backend API /schedule.
 */

import api from './api';
import type {
  ScheduleEvent,
  ScheduleEventListItem,
  EventParticipant,
  PaginatedEvents,
  EventFilters,
  EventCreateRequest,
  EventUpdateRequest,
  CalendarDay,
  ScheduleStats,
} from '@/types';

// =============================================================================
// Transformers (snake_case -> camelCase)
// =============================================================================

/**
 * Преобразовать участника из API формата.
 */
function transformParticipant(data: Record<string, unknown>): EventParticipant {
  return {
    id: data.id as number,
    eventId: data.event_id as number,
    userId: data.user_id as number,
    role: data.role as EventParticipant['role'],
    status: data.status as EventParticipant['status'],
    note: data.note as string | null,
    userName: data.user_name as string | null,
  };
}

/**
 * Преобразовать событие (полное) из API формата.
 */
function transformEvent(data: Record<string, unknown>): ScheduleEvent {
  return {
    id: data.id as number,
    title: data.title as string,
    description: data.description as string | null,
    eventType: data.event_type as ScheduleEvent['eventType'],
    status: data.status as ScheduleEvent['status'],
    eventDate: data.event_date as string,
    startTime: data.start_time as string,
    endTime: data.end_time as string | null,
    venue: data.venue as string | null,
    performanceId: data.performance_id as number | null,
    color: data.color as string | null,
    metadata: data.metadata as Record<string, unknown> | null,
    isPublic: data.is_public as boolean,
    isActive: data.is_active as boolean,
    theaterId: data.theater_id as number | null,
    createdAt: data.created_at as string,
    updatedAt: data.updated_at as string,
    performanceTitle: data.performance_title as string | null,
    participants: data.participants 
      ? (data.participants as Record<string, unknown>[]).map(transformParticipant)
      : [],
  };
}

/**
 * Преобразовать событие для списка из API формата.
 */
function transformEventListItem(data: Record<string, unknown>): ScheduleEventListItem {
  return {
    id: data.id as number,
    title: data.title as string,
    eventType: data.event_type as ScheduleEventListItem['eventType'],
    status: data.status as ScheduleEventListItem['status'],
    eventDate: data.event_date as string,
    startTime: data.start_time as string,
    endTime: data.end_time as string | null,
    venue: data.venue as string | null,
    color: data.color as string | null,
    isPublic: data.is_public as boolean,
    performanceId: data.performance_id as number | null,
    performanceTitle: data.performance_title as string | null,
    participantsCount: data.participants_count as number,
  };
}

/**
 * Преобразовать статистику из API формата.
 */
function transformStats(data: Record<string, unknown>): ScheduleStats {
  return {
    totalEvents: data.total_events as number,
    planned: data.planned as number,
    confirmed: data.confirmed as number,
    completed: data.completed as number,
    cancelled: data.cancelled as number,
    performancesCount: data.performances_count as number,
    rehearsalsCount: data.rehearsals_count as number,
    otherCount: data.other_count as number,
    upcomingEvents: data.upcoming_events as number,
  };
}

// =============================================================================
// Request Transformers (camelCase -> snake_case)
// =============================================================================

function toSnakeCase<T extends object>(data: T): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  
  for (const [key, value] of Object.entries(data)) {
    if (value === undefined) continue;
    
    // camelCase to snake_case
    const snakeKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
    result[snakeKey] = value;
  }
  
  return result;
}

// =============================================================================
// API Service
// =============================================================================

export const scheduleService = {
  // ===========================================================================
  // Events CRUD
  // ===========================================================================
  
  /**
   * Получить список событий с фильтрацией и пагинацией.
   */
  async getEvents(filters: EventFilters = {}): Promise<PaginatedEvents> {
    const params = new URLSearchParams();
    
    if (filters.search) params.append('search', filters.search);
    if (filters.dateFrom) params.append('date_from', filters.dateFrom);
    if (filters.dateTo) params.append('date_to', filters.dateTo);
    if (filters.eventType) params.append('event_type', filters.eventType);
    if (filters.status) params.append('status', filters.status);
    if (filters.performanceId) params.append('performance_id', String(filters.performanceId));
    if (filters.page) params.append('page', String(filters.page));
    if (filters.limit) params.append('limit', String(filters.limit));
    
    const response = await api.get(`/schedule?${params.toString()}`);
    const data = response.data;
    
    return {
      items: data.items.map(transformEventListItem),
      total: data.total,
      page: data.page,
      limit: data.limit,
      pages: data.pages,
    };
  },

  /**
   * Получить события за период (для календаря).
   * Возвращает массив событий вместо пагинированного ответа.
   */
  async getEventsForPeriod(params: {
    startDate: string;
    endDate: string;
    performanceId?: number;
    eventType?: string;
  }): Promise<ScheduleEvent[]> {
    const queryParams = new URLSearchParams();
    queryParams.append('date_from', params.startDate.split('T')[0]);
    queryParams.append('date_to', params.endDate.split('T')[0]);
    queryParams.append('limit', '100'); // Достаточно для месячного view
    
    if (params.performanceId) {
      queryParams.append('performance_id', String(params.performanceId));
    }
    if (params.eventType) {
      queryParams.append('event_type', params.eventType);
    }
    
    const response = await api.get(`/schedule?${queryParams.toString()}`);
    const data = response.data;
    
    // Преобразуем items из списочного формата в полный формат
    // Для календаря нам достаточно базовых полей
    return data.items.map((item: Record<string, unknown>) => ({
      id: item.id as number,
      title: item.title as string,
      description: null,
      eventType: item.event_type as ScheduleEvent['eventType'],
      status: item.status as ScheduleEvent['status'],
      eventDate: item.event_date as string,
      startTime: item.start_time as string,
      endTime: item.end_time as string | null,
      venue: item.venue as string | null,
      performanceId: item.performance_id as number | null,
      color: item.color as string | null,
      metadata: null,
      isPublic: item.is_public as boolean,
      isActive: true,
      theaterId: null,
      createdAt: '',
      updatedAt: '',
      performanceTitle: item.performance_title as string | null,
      participants: [],
    }));
  },

  /**
   * Получить предстоящие события.
   */
  async getUpcoming(days: number = 7, limit: number = 10): Promise<ScheduleEventListItem[]> {
    const params = new URLSearchParams({
      days: String(days),
      limit: String(limit),
    });
    
    const response = await api.get(`/schedule/upcoming?${params.toString()}`);
    return response.data.map(transformEventListItem);
  },

  /**
   * Получить событие по ID.
   */
  async getEvent(id: number): Promise<ScheduleEvent> {
    const response = await api.get(`/schedule/${id}`);
    return transformEvent(response.data);
  },

  /**
   * Создать событие.
   */
  async createEvent(data: EventCreateRequest): Promise<ScheduleEvent> {
    const response = await api.post('/schedule', toSnakeCase(data));
    return transformEvent(response.data);
  },

  /**
   * Обновить событие.
   */
  async updateEvent(id: number, data: EventUpdateRequest): Promise<ScheduleEvent> {
    const response = await api.patch(`/schedule/${id}`, toSnakeCase(data));
    return transformEvent(response.data);
  },

  /**
   * Удалить событие.
   */
  async deleteEvent(id: number): Promise<void> {
    await api.delete(`/schedule/${id}`);
  },

  // ===========================================================================
  // Status Actions
  // ===========================================================================

  /**
   * Подтвердить событие.
   */
  async confirmEvent(id: number): Promise<ScheduleEvent> {
    const response = await api.post(`/schedule/${id}/confirm`);
    return transformEvent(response.data);
  },

  /**
   * Начать событие.
   */
  async startEvent(id: number): Promise<ScheduleEvent> {
    const response = await api.post(`/schedule/${id}/start`);
    return transformEvent(response.data);
  },

  /**
   * Завершить событие.
   */
  async completeEvent(id: number): Promise<ScheduleEvent> {
    const response = await api.post(`/schedule/${id}/complete`);
    return transformEvent(response.data);
  },

  /**
   * Отменить событие.
   */
  async cancelEvent(id: number): Promise<ScheduleEvent> {
    const response = await api.post(`/schedule/${id}/cancel`);
    return transformEvent(response.data);
  },

  // ===========================================================================
  // Participants
  // ===========================================================================

  /**
   * Получить участников события.
   */
  async getParticipants(eventId: number): Promise<EventParticipant[]> {
    const response = await api.get(`/schedule/${eventId}/participants`);
    return response.data.map(transformParticipant);
  },

  /**
   * Добавить участника.
   */
  async addParticipant(
    eventId: number,
    data: { userId: number; role?: string; note?: string }
  ): Promise<EventParticipant> {
    const response = await api.post(`/schedule/${eventId}/participants`, toSnakeCase(data));
    return transformParticipant(response.data);
  },

  /**
   * Удалить участника.
   */
  async removeParticipant(participantId: number): Promise<void> {
    await api.delete(`/schedule/participants/${participantId}`);
  },

  /**
   * Ответить на приглашение.
   */
  async respondToEvent(
    eventId: number,
    status: 'confirmed' | 'declined' | 'tentative'
  ): Promise<EventParticipant> {
    const response = await api.post(`/schedule/${eventId}/respond?status=${status}`);
    return transformParticipant(response.data);
  },

  // ===========================================================================
  // Calendar
  // ===========================================================================

  /**
   * Получить календарь на месяц.
   */
  async getCalendar(year: number, month: number): Promise<CalendarDay[]> {
    const response = await api.get(`/schedule/calendar/${year}/${month}`);
    
    return response.data.map((day: Record<string, unknown>) => ({
      date: day.date as string,
      events: (day.events as Record<string, unknown>[]).map((e) => ({
        id: e.id as number,
        title: e.title as string,
        eventType: e.event_type as string,
        status: e.status as string,
        date: e.date as string,
        start: e.start as string,
        end: e.end as string | null,
        color: e.color as string | null,
        performanceId: e.performance_id as number | null,
      })),
    }));
  },

  // ===========================================================================
  // Stats
  // ===========================================================================

  /**
   * Получить статистику расписания.
   */
  async getStats(): Promise<ScheduleStats> {
    const response = await api.get('/schedule/stats/');
    return transformStats(response.data);
  },
};

export default scheduleService;

File: backend/app/services/schedule_service.py
"""
Сервис модуля расписания.

Бизнес-логика для работы с расписанием:
- CRUD операции с событиями
- Управление участниками
- Календарное представление
"""
from datetime import date, timedelta

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import NotFoundError, ValidationError
from app.models.schedule import (
    ScheduleEvent,
    EventParticipant,
    EventType,
    EventStatus,
    ParticipantStatus,
)
from app.repositories.schedule_repository import (
    ScheduleEventRepository,
    EventParticipantRepository,
)
from app.schemas.schedule import (
    EventCreate,
    EventUpdate,
    ParticipantCreate,
    ParticipantUpdate,
    ScheduleStats,
    CalendarEvent,
    CalendarDay,
)


class ScheduleService:
    """
    Сервис расписания.
    
    Управляет событиями, участниками и календарём.
    """
    
    def __init__(self, session: AsyncSession):
        self._session = session
        self._event_repo = ScheduleEventRepository(session)
        self._participant_repo = EventParticipantRepository(session)
    
    # =========================================================================
    # Events
    # =========================================================================
    
    async def get_events(
        self,
        search: str | None = None,
        date_from: date | None = None,
        date_to: date | None = None,
        event_type: EventType | None = None,
        status: EventStatus | None = None,
        performance_id: int | None = None,
        theater_id: int | None = None,
        skip: int = 0,
        limit: int = 20,
    ) -> tuple[list[ScheduleEvent], int]:
        """Получить список событий с фильтрацией."""
        events, total = await self._event_repo.search(
            search=search,
            date_from=date_from,
            date_to=date_to,
            event_type=event_type,
            status=status,
            performance_id=performance_id,
            theater_id=theater_id,
            skip=skip,
            limit=limit,
        )
        return list(events), total
    
    async def get_event(self, event_id: int) -> ScheduleEvent:
        """Получить событие по ID."""
        event = await self._event_repo.get_with_relations(event_id)
        if not event:
            raise NotFoundError(f"Событие с ID {event_id} не найдено")
        return event
    
    async def get_events_by_date(
        self,
        event_date: date,
        theater_id: int | None = None,
    ) -> list[ScheduleEvent]:
        """Получить события на дату."""
        events = await self._event_repo.get_by_date(event_date, theater_id)
        return list(events)
    
    async def get_events_by_range(
        self,
        date_from: date,
        date_to: date,
        theater_id: int | None = None,
        event_type: EventType | None = None,
    ) -> list[ScheduleEvent]:
        """Получить события за период."""
        events = await self._event_repo.get_by_date_range(
            date_from, date_to, theater_id, event_type
        )
        return list(events)
    
    async def get_upcoming(
        self,
        days: int = 7,
        theater_id: int | None = None,
        limit: int = 10,
    ) -> list[ScheduleEvent]:
        """Получить предстоящие события."""
        events = await self._event_repo.get_upcoming(days, theater_id, limit)
        return list(events)
    
    async def create_event(
        self,
        data: EventCreate,
        user_id: int,
        theater_id: int | None = None,
    ) -> ScheduleEvent:
        """Создать событие."""
        # Создаём событие
        event = ScheduleEvent(
            title=data.title,
            description=data.description,
            event_type=data.event_type,
            status=EventStatus.PLANNED,
            event_date=data.event_date,
            start_time=data.start_time,
            end_time=data.end_time,
            venue=data.venue,
            performance_id=data.performance_id,
            color=data.color,
            is_public=data.is_public,
            theater_id=theater_id,
            created_by_id=user_id,
            updated_by_id=user_id,
        )
        
        self._session.add(event)
        await self._session.flush()
        
        # Добавляем участников
        for p_data in data.participants:
            participant = EventParticipant(
                event_id=event.id,
                user_id=p_data.user_id,
                role=p_data.role,
                status=p_data.status,
                note=p_data.note,
            )
            self._session.add(participant)
        
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event.id)
    
    async def update_event(
        self,
        event_id: int,
        data: EventUpdate,
        user_id: int,
    ) -> ScheduleEvent:
        """Обновить событие."""
        await self.get_event(event_id)
        
        update_data = data.model_dump(exclude_unset=True)
        update_data["updated_by_id"] = user_id
        
        await self._event_repo.update(event_id, update_data)
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event_id)
    
    async def delete_event(self, event_id: int, user_id: int) -> bool:
        """Удалить событие (soft delete)."""
        await self.get_event(event_id)
        await self._event_repo.update(event_id, {
            "is_active": False,
            "updated_by_id": user_id,
        })
        await self._session.commit()
        return True
    
    # =========================================================================
    # Status Management
    # =========================================================================
    
    async def confirm_event(self, event_id: int, user_id: int) -> ScheduleEvent:
        """Подтвердить событие."""
        event = await self.get_event(event_id)
        
        if event.status not in [EventStatus.PLANNED]:
            raise ValidationError("Можно подтвердить только запланированное событие")
        
        await self._event_repo.update(event_id, {
            "status": EventStatus.CONFIRMED,
            "updated_by_id": user_id,
        })
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event_id)
    
    async def start_event(self, event_id: int, user_id: int) -> ScheduleEvent:
        """Начать событие."""
        event = await self.get_event(event_id)
        
        if event.status not in [EventStatus.PLANNED, EventStatus.CONFIRMED]:
            raise ValidationError("Невозможно начать событие в текущем статусе")
        
        await self._event_repo.update(event_id, {
            "status": EventStatus.IN_PROGRESS,
            "updated_by_id": user_id,
        })
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event_id)
    
    async def complete_event(self, event_id: int, user_id: int) -> ScheduleEvent:
        """Завершить событие."""
        event = await self.get_event(event_id)
        
        if event.status not in [EventStatus.IN_PROGRESS, EventStatus.CONFIRMED]:
            raise ValidationError("Невозможно завершить событие в текущем статусе")
        
        await self._event_repo.update(event_id, {
            "status": EventStatus.COMPLETED,
            "updated_by_id": user_id,
        })
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event_id)
    
    async def cancel_event(self, event_id: int, user_id: int) -> ScheduleEvent:
        """Отменить событие."""
        event = await self.get_event(event_id)
        
        if event.status == EventStatus.COMPLETED:
            raise ValidationError("Нельзя отменить завершённое событие")
        
        await self._event_repo.update(event_id, {
            "status": EventStatus.CANCELLED,
            "updated_by_id": user_id,
        })
        await self._session.commit()
        
        return await self._event_repo.get_with_relations(event_id)
    
    # =========================================================================
    # Participants
    # =========================================================================
    
    async def get_participants(self, event_id: int) -> list[EventParticipant]:
        """Получить участников события."""
        await self.get_event(event_id)
        participants = await self._participant_repo.get_by_event(event_id)
        return list(participants)
    
    async def add_participant(
        self,
        event_id: int,
        data: ParticipantCreate,
    ) -> EventParticipant:
        """Добавить участника."""
        await self.get_event(event_id)
        
        # Проверяем, нет ли уже такого участника
        existing = await self._participant_repo.get_by_user_and_event(
            data.user_id, event_id
        )
        if existing:
            raise ValidationError("Участник уже добавлен в событие")
        
        participant = EventParticipant(
            event_id=event_id,
            user_id=data.user_id,
            role=data.role,
            status=data.status,
            note=data.note,
        )
        
        self._session.add(participant)
        await self._session.commit()
        
        return await self._participant_repo.get_by_id(participant.id)
    
    async def update_participant(
        self,
        participant_id: int,
        data: ParticipantUpdate,
    ) -> EventParticipant:
        """Обновить участника."""
        participant = await self._participant_repo.get_by_id(participant_id)
        if not participant:
            raise NotFoundError(f"Участник с ID {participant_id} не найден")
        
        update_data = data.model_dump(exclude_unset=True)
        await self._participant_repo.update(participant_id, update_data)
        await self._session.commit()
        
        return await self._participant_repo.get_by_id(participant_id)
    
    async def remove_participant(self, participant_id: int) -> bool:
        """Удалить участника."""
        participant = await self._participant_repo.get_by_id(participant_id)
        if not participant:
            raise NotFoundError(f"Участник с ID {participant_id} не найден")
        
        await self._session.delete(participant)
        await self._session.commit()
        return True
    
    async def respond_to_event(
        self,
        event_id: int,
        user_id: int,
        status: ParticipantStatus,
    ) -> EventParticipant:
        """Ответить на приглашение."""
        participant = await self._participant_repo.get_by_user_and_event(
            user_id, event_id
        )
        if not participant:
            raise NotFoundError("Вы не являетесь участником этого события")
        
        await self._participant_repo.update(participant.id, {"status": status})
        await self._session.commit()
        
        return await self._participant_repo.get_by_id(participant.id)
    
    # =========================================================================
    # Calendar
    # =========================================================================
    
    async def get_calendar(
        self,
        year: int,
        month: int,
        theater_id: int | None = None,
    ) -> list[CalendarDay]:
        """Получить календарь на месяц."""
        # Определяем границы месяца
        first_day = date(year, month, 1)
        if month == 12:
            last_day = date(year + 1, 1, 1) - timedelta(days=1)
        else:
            last_day = date(year, month + 1, 1) - timedelta(days=1)
        
        # Получаем события
        events = await self._event_repo.get_by_date_range(
            first_day, last_day, theater_id
        )
        
        # Группируем по дням
        days_dict: dict[date, list[CalendarEvent]] = {}
        
        for event in events:
            cal_event = CalendarEvent(
                id=event.id,
                title=event.title,
                event_type=event.event_type,
                status=event.status,
                date=event.event_date,
                start=event.start_time,
                end=event.end_time,
                color=event.color,
                performance_id=event.performance_id,
            )
            
            if event.event_date not in days_dict:
                days_dict[event.event_date] = []
            days_dict[event.event_date].append(cal_event)
        
        # Формируем список дней
        calendar_days = []
        current = first_day
        while current <= last_day:
            calendar_days.append(CalendarDay(
                date=current,
                events=days_dict.get(current, []),
            ))
            current += timedelta(days=1)
        
        return calendar_days
    
    # =========================================================================
    # Statistics
    # =========================================================================
    
    async def get_stats(self, theater_id: int | None = None) -> ScheduleStats:
        """Получить статистику расписания."""
        stats = await self._event_repo.get_stats(theater_id)
        
        return ScheduleStats(
            total_events=stats["total_events"],
            planned=stats.get("planned", 0),
            confirmed=stats.get("confirmed", 0),
            completed=stats.get("completed", 0),
            cancelled=stats.get("cancelled", 0),
            performances_count=stats.get("performances_count", 0),
            rehearsals_count=stats.get("rehearsals_count", 0),
            other_count=stats.get("other_count", 0),
            upcoming_events=stats.get("upcoming_events", 0),
        )
